core/src/main/java/jenkins/util/SystemProperties.java

```diff
@@ -0,0 +1,253 @@
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * Centralizes calls to {@link System#getProperty()} and related calls.
+ * This allows us to get values not just from environment variables but also from
+ * the {@link ServletContext}, so properties like {@code hudson.DNSMultiCast.disabled}
+ * can be set in {@code context.xml} and the app server's boot script does not
+ * have to be changed.
+ *
+ * <p>This should be used to obtain hudson/jenkins "app"-level parameters
+ * (e.g. {@code hudson.DNSMultiCast.disabled}), but not for system parameters
+ * (e.g. {@code os.name}).
+ *
+ * <p>If you run multiple instances of Jenkins in the same virtual machine and wish
+ * to obtain properties from {@code context.xml}, make sure these Jenkins instances use
+ * different ClassLoaders. Tomcat, for example, does this automatically. If you do
+ * not use different ClassLoaders, the values of properties specified in
+ * {@code context.xml} is undefined.
+ *
+ * <p>Property access is logged on {@link Level#CONFIG}. Note that some properties
+ * may be accessed by Jenkins before logging is configured properly, so early access to
+ * some properties may not be logged.
+ *
+ * <p>While it looks like it on first glance, this cannot be mapped to {@link EnvVars},
+ * because {@link EnvVars} is only for build variables, not Jenkins itself variables.
+ *
+ * @author Johannes Ernst
+ * @since TODO
+ */
+public class SystemProperties {
+    /**
+     * The ServletContext to get the "init" parameters from.
+     */
+    @CheckForNull
+    private static ServletContext theContext;
+
+    /**
+     * Logger.
+     */
+    private static final Logger LOGGER = Logger.getLogger(SystemProperties.class.getName());
+
+    /**
+     * This class should never be instantiated.
+     */
+    private SystemProperties() {}
+
+    /**
+     * Gets the system property indicated by the specified key.
+     * This behaves just like {@link System#getProperty(java.lang.String)}, except that it
+     * also consults the {@link ServletContext}'s "init" parameters.
+     * 
+     * @param      key   the name of the system property.
+     * @return     the string value of the system property,
+     *             or {@code null} if there is no property with that key.
+     *
+     * @exception  NullPointerException if {@code key} is {@code null}.
+     * @exception  IllegalArgumentException if {@code key} is empty.
+     */
+    @CheckForNull
+    public static String getString(String key) {
+        String value = System.getProperty(key); // keep passing on any exceptions
+        if (value != null) {
+            if (LOGGER.isLoggable(Level.CONFIG)) {
+                LOGGER.log(Level.CONFIG, "Property (system): {0} => {1}", new Object[] {key, value});
+            }
+            return value;
+        }
+        
+        value = tryGetValueFromContext(key);
+        if (value != null) {
+            LOGGER.log(Level.CONFIG, "Property (context): {0} => {1}", new Object[]{key, value});
```

Comment:
This just introduces code fluff and is not needed.  None of the things being passed into the array require computation and an array creation is pretty optimized.
The code is not on the critical path so all this does is an uneeded if check...